README.md
# TextRich

A composite rich text component.

## Getting Started

### Screenshot

<img src="https://github.com/laoeg/TextRich/blob/main/img.png" alt="drawing" width="200"/>

### Code

```
TextRich(
  "TextRich example app",
  matchTexts: const ["exa"],
  matchStyles: const [TextStyle(color: Colors.red)],
),
TextRich(
  "TextRich example app",
  matchTexts: const ["exa", "a"],
  matchStyles: const [TextStyle(color: Colors.red)],
),
TextRich(
  "TextRich example app",
  matchTexts: const ["exa", "a"],
  matchStyles: const [TextStyle(color: Colors.red), TextStyle(color: Colors.blue)],
),
TextRich(
  "TextRich example app",
  matchTexts: const ["exa", "a"],
  matchStyles: const [TextStyle(color: Colors.red,fontStyle: FontStyle.italic,fontSize: 30,decoration: TextDecoration.underline), TextStyle(color: Colors.blue)],
  onTap: (value){
    print('Tap value :$value');
  },
  style: const TextStyle(color: Colors.black,fontSize: 16),
),
```

CHANGELOG.md
## 0.0.1

* TODO: Describe initial release.

## 0.0.2

* Improve documentation

## 0.0.3

* add Screenshot

## 0.0.4

* remove plugin_platform_interface

## 0.0.5

* update README

test/textrich_test.dart

import 'package:flutter_test/flutter_test.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';
import 'package:textrich/textrich.dart';
import 'package:textrich/textrich_method_channel.dart';
import 'package:textrich/textrich_platform_interface.dart';

class MockTextrichPlatform
    with MockPlatformInterfaceMixin
    implements TextrichPlatform {

  @override
  Future<String?> getPlatformVersion() => Future.value('42');
}

void main() {
  final TextrichPlatform initialPlatform = TextrichPlatform.instance;

  test('$MethodChannelTextrich is the default instance', () {
    expect(initialPlatform, isInstanceOf<MethodChannelTextrich>());
  });

  test('getPlatformVersion', () async {
    TextRich textrichPlugin = TextRich("");
    MockTextrichPlatform fakePlatform = MockTextrichPlatform();
    TextrichPlatform.instance = fakePlatform;

    // expect(await textrichPlugin.getPlatformVersion(), '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:textrich_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/textrich_method_channel_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('textrich');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

}


lib/textrich.dart
import 'dart:math';

import 'package:collection/collection.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

typedef TapCallBack = Function(String value);

class Node {
  final int start;
  final String element;
  final int end;
  final int styleIndex;

  Node(this.start, this.element, this.end, this.styleIndex);
}

class TextRich extends StatelessWidget {
  final String text;
  final TextStyle? style;
  final List<String>? matchTexts;
  final List<TextStyle>? matchStyles;
  final TapCallBack? onTap;

  const TextRich(this.text, {this.style, this.matchTexts, this.matchStyles, this.onTap, Key? key})
      : assert(
  (matchTexts == null && matchStyles == null) ||
      (matchTexts != null &&
          matchTexts.length > 0 &&
          matchStyles != null &&
          matchTexts.length >= matchStyles.length),
  "match text length must greater than match style length and all length must greater than 1"),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    List<TextSpan> textSpanChildren = [];
    int lastIndex = text.length;
    if (matchStyles != null && matchTexts != null) {
      ///1.获取所有的匹配字符串
      List<Node> allNodes = [];
      for (var i = 0; i < matchTexts!.length; i++) {
        var regExp = RegExp(matchTexts![i]);
        var matches = regExp.allMatches(text);
        var nodes = matches.map((e) {
          var element = e.group(0);
          return Node(e.start, element ?? "", e.end, i);
        }).toList();
        allNodes.addAll(nodes);
      }

      ///2.按在text中的下标从小到大排序
      var result = allNodes.sorted((c, n) {
        return c.start.compareTo(n.start);
      });


      for (int i1 = result.length - 1; i1 >= 0; i1--) {
        var node = result[i1];
        Node? lastNode = null;
        if(node.start>=lastIndex){

          // textSpanChildren.removeLast();
          // if(lastNode != null) {
          //   textSpanChildren.add(TextSpan(
          //       text: node.element.substring(0, node.start - lastIndex + 1), style: matchStyles![lastNode.styleIndex]));
          // }
          // var r = null;
          // if (this.onTap != null) {
          //   r = TapGestureRecognizer()
          //     ..onTap = () {
          //       this.onTap?.call(node.element);
          //     };
          // }
          // textSpanChildren.add(TextSpan(
          //     text: node.element.substring(node.start-lastIndex+1),
          //     style: matchStyles![node.styleIndex >= matchStyles!.length ? matchStyles!.length - 1 : node.styleIndex],
          //     recognizer: r));
          continue;
        }
        lastNode = node;
        var end = min(lastIndex, node.end);
        textSpanChildren.add(TextSpan(text: text.substring(end, lastIndex)));
        lastIndex = node.start;
        var r = null;
        if (this.onTap != null) {
          r = TapGestureRecognizer()
            ..onTap = () {
              this.onTap?.call(node.element);
            };
        }
        textSpanChildren.add(TextSpan(
            text: text.substring(node.start, end),
            style: matchStyles![node.styleIndex >= matchStyles!.length ? matchStyles!.length - 1 : node.styleIndex],
            recognizer: r));
      }

      ///3.合并相同元素，合并部分字符串重叠元素
      // List<Node> datas = [];
      // for (int j = 0; j < result.length; j++) {
      //   var node = result[j];
      //   if (datas.length == 0) {
      //     datas.add(node);
      //   } else {
      //     var current = datas.last;
      //     var next = node;
      //     if (next.start - current.end > 0) {
      //       datas.add(next);
      //     } else {
      //       if (current.element == next.element) {
      //         continue;
      //       } else {
      //         datas.removeLast();
      //         datas.add(Node(current.start, current.element + next.element.substring(current.end - next.start),
      //             next.end, next.styleIndex));
      //       }
      //     }
      //   }
      // }
      //
      // ///4.生成widget列表
      // for (var i = 0; i < datas.length; i++) {
      //   if (datas[i].element.isEmpty) {
      //     continue;
      //   }
      //   int index = text.indexOf(datas[i].element);
      //   textSpanChildren.add(TextSpan(text: text.substring(lastIndex, index)));
      //   lastIndex = index + datas[i].element.length;
      //   var r = null;
      //   if (this.onTap != null) {
      //     r = TapGestureRecognizer()
      //       ..onTap = () {
      //         this.onTap?.call(datas[i].element);
      //       };
      //   }
      //   textSpanChildren.add(TextSpan(
      //       text: datas[i].element,
      //       style: matchStyles![datas[i].styleIndex >= matchStyles!.length ? matchStyles!.length-1 : datas[i].styleIndex],
      //       recognizer: r));
      // }
    }
    if (lastIndex != 0) {
      textSpanChildren.add(TextSpan(text: text.substring(0, lastIndex)));
    }
    return Text.rich(
      TextSpan(children: textSpanChildren.reversed.toList()),
      style: style,
    );
  }
}


lib/textrich_platform_interface.dart
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

import 'textrich_method_channel.dart';

abstract class TextrichPlatform extends PlatformInterface {
  /// Constructs a TextrichPlatform.
  TextrichPlatform() : super(token: _token);

  static final Object _token = Object();

  static TextrichPlatform _instance = MethodChannelTextrich();

  /// The default instance of [TextrichPlatform] to use.
  ///
  /// Defaults to [MethodChannelTextrich].
  static TextrichPlatform get instance => _instance;

  /// Platform-specific implementations should set this with their own
  /// platform-specific class that extends [TextrichPlatform] when
  /// they register themselves.
  static set instance(TextrichPlatform instance) {
    PlatformInterface.verifyToken(instance, _token);
    _instance = instance;
  }

  Future<String?> getPlatformVersion() {
    throw UnimplementedError('platformVersion() has not been implemented.');
  }
}


android/src/main/java/com/example/textrich/TextrichPlugin.java
package com.example.textrich;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** TextrichPlugin */
public class TextrichPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "textrich");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:textrich/textrich.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
  }


  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('TextRich example app'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            children: [
              TextRich(
                "TextRich example app",
                matchTexts: const ["exa"],
                matchStyles: const [TextStyle(color: Colors.red)],
              ),
              TextRich(
                "TextRich example app",
                matchTexts: const ["exa", "a"],
                matchStyles: const [TextStyle(color: Colors.red)],
              ),
              TextRich(
                "TextRich example app",
                matchTexts: const ["exa", "a"],
                matchStyles: const [TextStyle(color: Colors.red), TextStyle(color: Colors.blue)],
              ),
              TextRich(
                "TextRich example app",
                matchTexts: const ["exa", "a"],
                matchStyles: const [TextStyle(color: Colors.red,fontStyle: FontStyle.italic,fontSize: 30,decoration: TextDecoration.underline), TextStyle(color: Colors.blue)],
                onTap: (value){
                  print('Tap value :$value');
                },
                style: const TextStyle(color: Colors.black,fontSize: 16),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
